apis/python/src/tiledb/vector_search/object_readers/directory_reader.py

```diff
@@ -8,6 +18,223 @@
+                        results.append(file_uri)
+
+        # Add one trailing slash to search_uri
+        search_uri = search_uri.rstrip("/") + "/"
+        results = []
+        find(
+            search_uri,
+            include=include,
+            exclude=exclude,
+            suffixes=suffixes,
+            max_files=max_files,
+        )
+        if max_files is not None:
+            return results[:max_files]
+        else:
+            return results
+
+
+def find_uris_tiledb_group(
+    search_uri: str,
+    config: Optional[Mapping[str, Any]] = None,
+    include: str = "*",
+    exclude: Sequence[str] = ["[.]*", "*/[.]*"],
+    suffixes: Optional[Sequence[str]] = None,
+    max_files: Optional[int] = None,
+    verbose: bool = False,
+) -> Sequence[str]:
+    with tiledb.scope_ctx(config):
+
+        def find(
+            uri: str,
+            path: str,
+            include: str = "*",
+            exclude: Sequence[str] = (),
+            suffixes: Optional[Sequence[str]] = None,
+            max_files: Optional[int] = None,
+        ):
+            # Stop if we have found max_count files
+            if max_files is not None and len(results) >= max_files:
+                return
+            obj_type = tiledb.object_type(uri)
+            if obj_type == "array":
+                match = match_uri(
+                    search_uri="/",
+                    file_uri=path,
+                    include=include,
+                    exclude=exclude,
+                    suffixes=suffixes,
+                )
+                if match:
+                    results.append(uri)
+            elif obj_type == "group":
+                with tiledb.Group(uri) as grp:
+                    for child in list(grp):
+                        find(
+                            uri=child.uri,
+                            path=f"{path}/{child.name}",
+                            include=include,
+                            exclude=exclude,
+                            suffixes=suffixes,
+                            max_files=max_files,
+                        )
+
+        results = []
+        find(
+            uri=search_uri,
+            path="",
+            include=include,
+            exclude=exclude,
+            suffixes=suffixes,
+            max_files=max_files,
+        )
+        if max_files is not None:
+            return results[:max_files]
+        else:
+            return results
+
+
+def find_uris_aws(
+    search_uri: str,
+    config: Optional[Mapping[str, Any]] = None,
+    include: str = "*",
+    exclude: Sequence[str] = ["[.]*", "*/[.]*"],
+    suffixes: Optional[Sequence[str]] = None,
+    max_files: Optional[int] = None,
+    verbose: bool = False,
+) -> Sequence[str]:
+    from urllib.parse import urlparse
+
+    try:
+        import boto3
+    except ImportError:
+        raise ImportError(
+            "Could not import boto3 python package. "
+            "Please install it with `pip install boto3`."
+        )
```

Comment:
(Curious) This is nice, but, curious, why do it only for `boto3`? Is it because it's a more uncommon package? Or something else I'm missing here?

Reply:
This function is not currently utilized for s3. We could reuse the same import pattern for libraries that the readers require but are not a dependency of the vector search module.